hal/src/common/dmac/transfer.rs

```diff
@@ -0,0 +1,864 @@
+        let descaddr = if circular_xfer {
+            // SAFETY: This is safe as we are only reading the descriptor's address,
+            // and not actually writing any data to it. We also assume the descriptor
+            // will never be moved.
+            unsafe {
+                DESCRIPTOR_SECTION.as_ptr() as u32
+                    + mem::size_of::<DmacDescriptor>() as u32 * (ID as u32)
+            }
+        } else {
+            0
+        };
+
+        let src = &mut config.source;
+        let dst = &mut config.dest;
+
+        let xfer_descriptor = DmacDescriptor {
+            // Next descriptor address:  0x0 terminates the transaction (no linked list),
+            // any other address points to the next block descriptor
+            descaddr,
+            // Source address: address of the last beat transfer source in block
+            srcaddr: src.to_dma_ptr() as u32,
+            // Destination address: address of the last beat transfer destination in block
+            dstaddr: dst.to_dma_ptr() as u32,
+            // Block transfer count: number of beats in block transfer
+            btcnt: config.length,
+            // Block transfer control: Datasheet  section 19.8.2.1 p.329
+            btctrl: ((src.incrementing() as u16) << 10) // Increment source address?
+                | ((dst.incrementing() as u16) << 11)   // Tncrement dest address?
+                | ((B::to_beatsize() as u16) << 8)      // Beatsize
+                | (1 << 0), // Validate descriptor
+        };
+
+        // SAFETY this is safe as long as we ONLY write to the descriptor
+        // belonging to OUR channel. We assume this is the only place
+        // in the entire library that this section or the array
+        // will be written to.
+        unsafe {
+            DESCRIPTOR_SECTION[ID as usize] = xfer_descriptor;
+        }
+
+        UnsafeTransfer {
+            chan,
+            config,
+            payload,
+        }
+    }
+
+    /// Begin DMA transfer. If [TriggerSource::DISABLE](TriggerSource::DISABLE)
+    /// is used, a sowftware trigger will be issued to the DMA channel to
+    /// launch the transfer. Is is therefore not necessary, in most cases,
+    /// to manually issue a software trigger to the channel.
+    pub unsafe fn begin(
+        self,
+        dmac: &mut DmaController,
+        trig_src: TriggerSource,
+        trig_act: TriggerAction,
+    ) -> UnsafeTransfer<B, P, Src, Dst, Busy, ID> {
+        // Compiler fence to prevent the compiler from re-ordering read/write operations
+        // beyond this fence.
+        // (see https://docs.rust-embedded.org/embedonomicon/dma.html#compiler-misoptimizations)
+        // (see https://docs.rust-embedded.org/embedonomicon/dma.html#compiler-misoptimizations)
+        // TODO change this to an atomic::fence since SAMD5x/SAME5x feature a
+        // data/instruction cache?
+        atomic::compiler_fence(atomic::Ordering::Release); //  ▲
```

Comment:
Is an `atomic::compiler_fence` sufficient here? [This](https://docs.rust-embedded.org/embedonomicon/dma.html#compiler-misoptimizations) seems to suggest so. Although since the SAMD51/SAME5x have a data/instruction cache, I'm not so sure if an `atomic::fence` would be more appropriate.

Reply:
Considering the following comment:
> In the case of Cortex-M7 cores you'll need memory barriers (DMB/DSB) if you are using the data cache (DCache), unless you manually invalidate the buffer used by the DMA. Even with the data cache disabled, memory barriers might still be required to avoid reordering in the store buffer.

I feel like it would be appropriate to use `atomic::fence` for those cores. Probably fine to just change to `atomic::fence` for everyone, or you could use one of the "-min" features to choose